src/lib/string_utils.ads

```diff
@@ -0,0 +1,56 @@
+with System;
+with System.Storage_Elements;
+use all type System.Address;
+use all type System.Storage_Elements.Integer_Address;
+
+package String_Utils
+   with SPARK_Mode
+is
+   pragma Preelaborate;
+
+   function Length (
+                    C_Str : System.Address;
+                    Max_Length : Natural := Natural'Last
+                   ) return Integer
+     with
+       Post => Length'Result <= Max_Length,
+       Contract_Cases =>
+         (C_Str = System.Null_Address  => Length'Result <= 0,
+          C_Str /= System.Null_Address => Length'Result >= 0);
```

Comment:
Wouldn't it make sense to let `Length` return `-1` in case of a null pointer? Although it is not strictly necessary in this case, it could be helpful in future applications of this function.

Reply:
I'm not sure if a distinction is necessary here. In both cases you should't work on the string. I'd rather let the nullpointer check to the user of this function and would not encourage to use this function to check for null pointers explicitly.

Reply:
Is there a reason not to use a precondition that ensures no null pointer is passed to the function?

Reply:
Currently I also use it to save a check in the string conversion function (and I need it to initialize the array length so it needs to be able to handle null pointers). When length returns something greater than null I can prove that the address provided to length was not null. So I can check only on the length and do not need to check for a null pointer.

Reply:
I don't think I understand. Can you point me to the code you're referring to?

Reply:
First of all the `Length` function is meant to be used in code that we cannot prove or that gets unchecked values from code we have no control over, so it should be able to handle a passed null pointer. So a precondition requiring a valid pointer will fail in this case.

Secondly, if the `Length` function returns any value greater than 0 the pointer passed cannot be null (how else could it have calculated a length then?). On the other side this means that if there has been a null pointer the return value must be zero (but not the other way round, it can be zero even if the pointer is valid).  This is what the contract cases express.

The important part comes at https://github.com/jklmnn/ada-runtime/blob/issue_9_pr/src/lib/string_utils.adb#L38. To get the required length of `Str` I need the length of the string behind the pointer first. To do this at the function declaration part (without a separate declare block) `Length` needs to be able to handle null pointers (as `Convert_To_Ada` is also meant to get untrusted input.
The only case that requires us to process anything is when `L` is greater than zero. The `Get_Char` and `Incr` both have a percondition that they do not accept null pointers. The easy way would be a further check at the beginning of the function (e.g. `if Cursor /= Null_Pointer and then L > 0 then`) but due to the contract cases we already have proven that `Cursor /= Null_Pointer` is true by the fact that `L >0` is also true.

So this whole thing is basically a short cut that reuses the information generated by `Length` instead of rechecking it. I hope this was more understandable.